pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,224 @@
+package azurensg
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+	machinev1beta1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	maoclient "github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	azureproviderv1beta1 "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
+
+	"github.com/Azure/ARO-RP/pkg/api/admin"
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/util/aad"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+)
+
+func (r *AzureNSGReconciler) reconcileSubnetNSG(ctx context.Context) error {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	masterSubnet, masterResourceGroup, err := r.masterParameters(ctx, r.maocli, instance.Spec.VnetID)
+	if err != nil {
+		return err
+	}
+
+	//Get endpoints from operator
+	azEnv, err := azure.EnvironmentFromName(instance.Spec.AZEnvironment)
+	if err != nil {
+		return err
+	}
+	//Grab azure-credentials from secret
+	credentials, err := azCredentials(ctx, r.kubernetescli)
+	if err != nil {
+		return err
+	}
+	//create service principal token from azure-credentials
+	token, err := aad.GetToken(ctx, r.log, string(credentials.clientID), string(credentials.clientSecret), string(credentials.tenantID), azEnv.ActiveDirectoryEndpoint, azEnv.ResourceManagerEndpoint)
+	if err != nil {
+		return err
+	}
+	//create refreshable authorizer from token
+	authorizer, err := newAuthorizer(token)
+	if err != nil {
+		return err
+	}
+	resource, err := azure.ParseResourceID(instance.Spec.ResourceID)
+	if err != nil {
+		return err
+	}
+
+	vnetResource, err := azure.ParseResourceID(instance.Spec.VnetID)
+	if err != nil {
+		return err
+	}
+
+	workerSubnets, err := r.workerSubnets(ctx, r.maocli)
+	if err != nil {
+		return err
+	}
+
+	subnetsClient := network.NewSubnetsClient(&azEnv, resource.SubscriptionID, authorizer)
+	switch admin.ArchitectureVersion(instance.Spec.ArchitectureVersion) {
+	case admin.ArchitectureVersionV1: // int == 0
+		//master subnet
+		v, err := r.isSubnetNSGValid(ctx, masterSubnet, vnetResource, resource, masterResourceGroup, subnetsClient, instance.Spec.InfraID+"-controlplane-nsg")
+		if err != nil {
+			return err
+		}
```

Comment:
What is stopper here to use dynamic validator `ValidateSubnets` function for this?
In addition please check `subnetutil.NetworkSecurityGroupID` to help fixups

Reply:
Dynamic validator relies on the full oc document, and I don't want to pass it all and worry about sanitizing it.
Instead I changed my code to use the same underlying code which generates the correct values as validate for NSGs so that there's one source of truth.